crates/core/tedge_actors/src/builders.rs

```diff
@@ -0,0 +1,232 @@
+use crate::mpsc;
+use crate::ClientId;
+use crate::ConcurrentServiceMessageBox;
+use crate::DynSender;
+use crate::KeyedSender;
+use crate::Message;
+use crate::NullSender;
+use crate::RequestResponseHandler;
+use crate::RuntimeError;
+use crate::RuntimeHandle;
+use crate::Sender;
+use crate::SenderVec;
+use crate::ServiceMessageBox;
+use crate::SimpleMessageBox;
+use async_trait::async_trait;
+use std::convert::Infallible;
+
+/// Materialize an actor instance under construction
+///
+/// Such an instance is:
+/// 1. built from some actor configuration
+/// 2. connected to other peers
+/// 3. eventually spawned into an actor.
+#[async_trait]
+pub trait ActorBuilder {
+    /// Build and spawn the actor
+    async fn spawn(self, runtime: &mut RuntimeHandle) -> Result<(), RuntimeError>;
+}
+
+/// Builder of `T`
+pub trait Builder<T>: Sized {
+    type Error: std::error::Error;
+
+    /// Builds the entity or returns an error
+    fn try_build(self) -> Result<T, Self::Error>;
+
+    /// Builds the entity or panics
+    fn build(self) -> T {
+        self.try_build().unwrap()
+    }
+}
+
+/// A trait to connect a message box under-construction to peer messages boxes
+pub trait MessageBoxConnector<Request: Message, Response: Message, Config: Default> {
+    /// Connect a peer message box to the message box under construction
+    fn connect_with(&mut self, peer: &mut impl MessageBoxPort<Request, Response>, config: Config);
+
+    /// Use the default config to connect a peer message box
+    fn connect(&mut self, peer: &mut impl MessageBoxPort<Request, Response>) {
+        self.connect_with(peer, Config::default())
+    }
+
+    /// Return a new request-response handle connected to this message box
+    fn new_handle(&mut self, client_name: &str) -> RequestResponseHandler<Request, Response> {
+        let mut port = RequestResponseHandlerBuilder::new(client_name, 1);
+        self.connect(&mut port);
+        port.build()
+    }
+}
+
+/// A connection port to connect a message box under-connection to another box
+pub trait MessageBoxPort<Request: Message, Response: Message> {
```

Comment:
Any idea to improve this name?

* `MessageBoxSocket`, since this a way to plug a peer actor mailbox to this message box?
* `ClientConnector`, matching the name `ServiceConnector`, to stress this is to connect a client to a service?
* `ClientSocket` ?

Reply:
This trait would need more documentation to better convey its purpose as it's not that straight-forward even after looking at the concrete implementations.

Reply:
Agree, the documentation must crucially be improved. => I will do that.

Reply:
We will rename this trait as `ActorPlug`.

Reply:
We are still looking for a better name.

* `ServiceConsumer`  or `ServiceClient` or `ServiceUser` might be good candidates as this trait defines the requirements of an actor for some service provided by another actor.
* An alternative is also to deprecate this trait and use a combination of `MessageSource + MessageSink` .

This will be addressed as a follow-up task.